# Welcome to Cerberus
Python code for versatile Functional Ontology Assignments for Metagenomes via Hidden Markov Model (HMM) searching with environmental focus of shotgun metaomics data

Input formats:
-----
- From any NextGen sequencing technology (from Illumina, Pacbio, Oxford Nanopore)
- type 1 raw reads (.fastq format)
- type 2 nucleotide fasta (.fasta, .fa, .fna, .ffn format), assembled raw reads into contigs
- type 3 protein fasta (.faa format), assembled contigs which genes are converted to amino acid sequence 

Table of Contents
-------

Installing Cerberus: 
-------
- needs a pip installer
`pip install cerberus`

- needs a conda installer
`conda install cerberus`

- needs source installer (github)
```
git clone https://github.com/raw-lab/cerberus.git
cd cerberus
python cerberus-setup.py
```

0. setup.py
-------
looks okay. I donâ€™t like the folder on the desktop.

## OSF link to FOAM databases </br>
https://osf.io/5ba2v/

- check  Install_Dependencies.py and Install_osf_Files.py 
