MetaCerberus Options
======================

.. important:: 
   If the MetaCerberus environment is not used, make sure the dependencies are in PATH or specified in the config file.

- Run ``metacerberus.py`` with the options required for your project.

Usage of ``metacerberus.py``: 

.. Note::
   The following are different options/arguments to modify the execution of MetaCerberus.


**Setup arguments:**
~~~~~~~~~~~~~~~~~~~~~~~~~
+-----------------+-------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+----------------------------+-----------------------------------------------+
|Argument/Option  | Function [Default]                                                                                                                        | Usage Format                   | Accepted format            |            Example (Type as one line)         |
+-----------------+-------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+----------------------------+-----------------------------------------------+
| ``--setup``     |  Setup additional dependencies [False]                                                                                                    |  ``--setup``                   |  N/A                       | ``metacerberus.py --setup``                   |  
+-----------------+-------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+----------------------------+-----------------------------------------------+
| ``--update``    | Update downloaded databases [False]                                                                                                       | ``--update``                   | N/A                        | ``metacerberus.py --update``                  |
+-----------------+-------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+----------------------------+-----------------------------------------------+
| ``--list-db``   | List available and downloaded databases [False]                                                                                           | ``--list-db``                  | N/A                        | ``metacerberus.py --list-db``                 | 
+-----------------+-------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+----------------------------+-----------------------------------------------+
| ``--download``  | Downloads selected HMMs. Use the option ``--list-db`` for a list of available databases, default is to download all available databases   | ``--download [DOWNLOAD ...]``  | ``--download [.HMM FILE]`` | ``--download path/to/example/directory.hmm``  |
+-----------------+-------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+----------------------------+-----------------------------------------------+
| ``--uninstall`` | Remove downloaded databases and FragGeneScan+ [False]                                                                                     | ``--uninstall``                | N/A                        | ``metacerberus.py --uninstall``               |
+-----------------+-------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+----------------------------+-----------------------------------------------+

**Input File Arguments:**
----------------------------

.. important:: 
  **At least one** sequence is required.

  **Accepted formats**: [.fastq, .fq, .fasta, .fa, .fna, .ffn, .faa]

   Example:
 
   - ``metacerberus.py --prodigal file1.fasta`` 
   - ``metacerberus.py --config file.config``

   If a sequence is given in [.fastq, .fq] format, one of ``--nanopore``, ``--illumina``, or ``--pacbio`` **is required.**:

   **Option format interpretation:**
   
   - ``--setup`` = accepts no additional options

   - ``--download DOWNLOAD`` = accepts one option, (represented by capitalized command 'DOWNLOAD')

   - ``--fraggenescan FRAGGENESCAN [FRAGGENESCAN...]`` = accepts one or greater options (represented by capitalized commands)

+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+              
|Argument/Option              | Function                                                                                                                                | Usage Format                                                            | Accepted format     | # Options Accepted | Example (Type as one line)                                                                |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``-c`` or ``--config``      | Path to config file, command line takes priority                                                                                        | ``-c CONFIG`` or ``--config CONFIG``                                    | Path to config file | 1                  | ``metacerberus.py -c path/to/config/file``                                                |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+ 
| ``--prodigal``              | Prokaryote nucleotide sequence (includes microbes, bacteriophage)                                                                       | ``--prodigal PRODIGAL [PRODIGAL ...]``                                  | Sequence file       | =>1                | ``metacerberus.py --prodigal FILE1 FILE2...``                                             |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--fraggenescan``          | Eukaryote nucleotide sequence (includes other viruses, works all around for everything)                                                 | ``--fraggenescan FRAGGENESCAN [FRAGGENESCAN ...]``                      | Sequence file       | =>1                | ``metacerberus.py --fraggenescan FILE1 FILE2...``                                         |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--super``                 | Runs sequence in **both** `--prodigal` and `--fraggenescan` modes                                                                       | ``--super SUPER [SUPER ...]``                                           | Sequence file       | =>1                | ``metacerberus.py --super FILE1 FILE2...``                                                |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--prodigalgv``            | Giant virus nucleotide sequence                                                                                                         | ``--prodigalgv PRODIGALGV [PRODIGALGV ...]``                            | Sequence file       | =>1                | ``metacerberus.py --prodigalgv FILE1 FILE2...``                                           |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--phanotate``             | Phage sequence                                                                                                                          | ``--phanotate PHANOTATE [PHANOTATE ...]``                               | Sequence file       | =>1                | ``metacerberus.py --phanotate  FILE1 FILE2...``                                           | 
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--protein`` or ``--amino``| Protein Amino Acid sequence                                                                                                             | ``--protein PROTEIN [PROTEIN ...]`` or ``--amino PROTEIN [PROTEIN ...]``| Sequence file       | =>1                | ``metacerberus.py --protein FILE1 FILE2...`` or ``metacerberus.py --amino FILE1 FILE2...``| 
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--hmmer-tsv``             | Annotations tsv file from HMMER (experimental)                                                                                          | ``--hmmer-tsv HMMER_TSV [HMMER_TSV ...]``                               | Sequence file       | =>1                | ``metacerberus.py --hmmer-tsv FILE1 FILE2...``                                            |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--class``                 | path to a tsv file which has class information for the samples. If this file is included, scripts will be included to run Pathview in R | ``--class CLASS``                                                       | Path to TSV file    | 1                  | ``metacerberus.py --class TSV_FILE1``                                                     |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--illumina``              | Specifies that the given FASTQ files are from Illumina                                                                                  | ``--illumina``                                                          | N/A                 | N/A                | ``metacerberus.py --illumina``                                                            |
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--nanopore``              | Specifies that the given FASTQ files are from Nanopore                                                                                  | ``--nanopore``                                                          | N/A                 | N/A                | ``metacerberus.py --nanopore``                                                            |  
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+
| ``--pacbio``                | Specifies that the given FASTQ files are from PacBio                                                                                    | ``--pacbio``                                                            | N/A                 | N/A                | ``metacerberus.py --pacbio``                                                              |  
+-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+---------------------+--------------------+-------------------------------------------------------------------------------------------+

**Output options:**
------------------------
+-----------------+-------------------------------------------------------------------------------------------------------------+----------------------+----------------------------+--------------------+---------------------------------------+
| Argument/Option | Function [DEFAULT]                                                                                          | Usage Format         | Accepted format            | # Options Accepted | Example (Type as one line)            |
+-----------------+-------------------------------------------------------------------------------------------------------------+----------------------+----------------------------+--------------------+---------------------------------------+
| ``--dir-out``   | path to output directory, defaults to "results-metacerberus" in current directory. [./results-`]            | ``--dir-out DIR_OUT``| output file path           | 1                  | ``--dir-out path/to/output/file``     |
+-----------------+-------------------------------------------------------------------------------------------------------------+----------------------+----------------------------+--------------------+---------------------------------------+
| ``--replace``   | Flag to replace existing files. [False]                                                                     | ``--replace``        | ``metacerberus.py`` option | N/A                | ``metacerberus.py --replace``         |
+-----------------+-------------------------------------------------------------------------------------------------------------+----------------------+----------------------------+--------------------+---------------------------------------+
| ``--keep``      | Flag to keep temporary files. [False]                                                                       | ``--keep``           | ``metacerberus.py`` option | N/A                | ``metacerberus.py --keep``            | 
+-----------------+-------------------------------------------------------------------------------------------------------------+----------------------+----------------------------+--------------------+---------------------------------------+
| ``--tmpdir``    | Temp directory for RAY (experimental) [system tmp dir]                                                      | ``--tmpdir TMPDIR``  | ``metacerberus.py`` option | 1                  | ``metacerberus.py --tmpdir TEMPFILE1``|
+-----------------+-------------------------------------------------------------------------------------------------------------+----------------------+----------------------------+--------------------+---------------------------------------+

**Database options:**
-------------------------
+----------------+-----------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+------------------------------------------------------+
|Argument/Option | Function [DEFAULT]                                                                                        | Usage Format            | Accepted format            | # Options Accepted | Example (Type as one line)                           |
+----------------+-----------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+------------------------------------------------------+
| ``--hmm``      | A list of databases for HMMER. Use the option ``--list-db`` for a list of available databases [KOFam_all] | ``--hmm HMM [HMM ...]`` | ``metacerberus.py`` option | =>1                | ``metacerberus.py --hmm DATABASE1 DATABASE2...``     |
+----------------+-----------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+------------------------------------------------------+
| ``--db-path``  | Path to folder of databases [Default: under the library path of metacerberus]                             | ``--db-path DB_PATH``   | path to databases folder   | 1                  | ``--db-path path/to/databases/folder``               | 
+----------------+-----------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+------------------------------------------------------+

**Optional Arguments:** 
--------------------------- 
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
|Argument/Option          | Function [DEFAULT]                                                                                                        | Usage Format            | Accepted format            | # Options Accepted | Example (Type as one line)                         |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--meta``              | Metagenomic nucleotide sequences **(for prodigal)** [False]                                                               | ``--meta``              | ``metacerberus.py`` option | N/A                | ``metacerberus.py --meta``                         |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--scaffolds``         | Sequences are treated as scaffolds [False]                                                                                | ``--scaffolds``         | ``metacerberus.py`` option | N/A                | ``metacerberus.py --scaffolds``                    |  
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--minscore``          | Score cutoff for parsing HMMER results [60]                                                                               | ``--minscore MINSCORE`` | whole integer value        | 1                  | ``metacerberus.py --minscore 50``                  |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--evalue``            | E-value cutoff for parsing HMMER results [1e-09]                                                                          | ``--evalue EVALUE``     | E-value                    | 1                  | ``metacerberus.py --evalue [E-value]``             | 
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--skip-decon``        | Skip decontamination step. [False]                                                                                        | ``--skip-decon``        | ``metacerberus.py`` option | N/A                | ``metacerberus.py --skip-decon``                   |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--skip-pca``          | Skip PCA. [False]                                                                                                         | ``--skip-pca``          | ``metacerberus.py`` option | N/A                | ``metacerberus.py --skip-pca``                     | 
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--cpus``              | Number of CPUs to use per task. System will try to detect available CPUs if not specified [Auto Detect]                   | ``--cpus CPUS``         | whole integer value        | 1                  | ``metacerberus.py --cpus 16``                      |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--chunker``           | Split files into smaller chunks, in Megabytes [Disabled by default]                                                       | ``--chunker CHUNKER``   | whole integer value        | 1                  | ``metacerberus.py --chunker 300``                  |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--grouped``           | Group multiple fasta files into a single file before processing. When used with `--chunker` (see above) can improve speed | ``--grouped``           | ``metacerberus.py`` option | N/A                | ``metacerberus.py --grouped``                      | 
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--version`` or ``-v`` | show the version number and exit                                                                                          | ``--version`` or ``-v`` | ``metacerberus.py`` option | N/A                | ``metacerberus.py --version``                      |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``-h`` or ``--help``    | show this help message and exit                                                                                           | ``-h`` or ``--help``    | ``metacerberus.py`` option | N/A                | ``metacerberus.py -h``                             |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--adapters``          | FASTA File containing adapter sequences for trimming                                                                      | ``--adapters ADAPTERS`` | FASTA file                 | 1                  | ``metacerberus.py --adapters /path/to/FASTA/file`` |   
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+
| ``--qc_seq``            | FASTA File containing control sequences for decontamination                                                               | ``--qc_seq QC_SEQ``     | FASTA file                 | 1                  | ``metacerberus.py --qc_seq /path/to/FASTA/file``   |
+-------------------------+---------------------------------------------------------------------------------------------------------------------------+-------------------------+----------------------------+--------------------+----------------------------------------------------+

.. note::
   Arguments/options that start with ``--`` can also be set in a config file (specified via ``-c``). Config file syntax allows: key=value, flag=true, stuff=[a,b,c] (for details, see `syntax`_. In general, **command-line values override config file values which override defaults.**
.. _syntax: https://goo.gl/R74nmi
